### PR TITLE
Show nodes with respect to the equator

### LINKED ISSUE #1841: Show ☊ and ☋ with respect to the equator in body-centred inertial frames
This came up during the design discussion in #1840.

This is not quite as simple as changing the conditions for [prediction nodes](https://github.com/mockingbirdnest/Principia/blob/2018051512-Darboux/ksp_plugin_adapter/ksp_plugin_adapter.cs#L1916-L1919) and [flight plan nodes](https://github.com/mockingbirdnest/Principia/blob/2018051512-Darboux/ksp_plugin_adapter/ksp_plugin_adapter.cs#L2000-L2003), since the xy plane for the body-centred inertial frames is currently the global reference plane (Earth's equator in RSS, Kerbin's in stock) rather than the plane of the equator of the relevant body.

Perhaps it would be easiest to change the definition of [`BodyCentredNonRotatingDynamicFrame`](https://github.com/mockingbirdnest/Principia/blob/2018051512-Darboux/ksp_plugin_adapter/ksp_plugin_adapter.cs#L2000-L2003) so that it uses the same z axis as [`BodySurfaceDynamicFrame`](https://github.com/mockingbirdnest/Principia/blob/2018051512-Darboux/physics/body_surface_dynamic_frame.hpp#L33-L36), making the y axis the node Q rather than making the x axis the point B from figure 1 of [the 2009 report of the IAU WGCCRE](https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2009reprint.pdf).

Note that Q≠`♈︎` for Earth, since α₀=0; this motivates making the y axis the node Q, rather than the x axis, so that the x axis points towards `♈︎` for Earth (in other words, `BodyCentredNonRotatingDynamicFrame` for Earth would be the same as `Barycentric` in RSS).

Addendum: it seems GitHub ignores VS-15 on U+2648 ARIES unless it is within a code block.

### PR DESCRIPTION
**Pull request by eggrobin**
This fixes #1841.

### CODE REVIEW

File: ksp_plugin/plugin.cpp [hunk 1]
```diff
@@ -937,18 +941,75 @@ void Plugin::ComputeAndRenderNodes(
...
+
+  std::optional<DiscreteTrajectory<Navigation>> filtered_ascending_trajectory;
+  std::optional<DiscreteTrajectory<Navigation>> filtered_descending_trajectory;
+
+  auto const cast_plotting_frame = dynamic_cast<
+      BodyCentredNonRotatingDynamicFrame<Barycentric, Navigation> const*>(
+      &*renderer_->GetPlottingFrame());
+  if (cast_plotting_frame != nullptr) {
+    auto const& plotting_frame = *cast_plotting_frame;
+    auto const centre = dynamic_cast_not_null<RotatingBody<Barycentric> const*>(
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Do `*dynamic_cast_not_null` and you'll avoid all the `->` below.

File: ksp_plugin/plugin.cpp [hunk 2]
```diff
@@ -937,18 +941,75 @@ void Plugin::ComputeAndRenderNodes(
                Vector<double, Navigation>({0, 0, 1}),
                ascending_trajectory,
                descending_trajectory);
-  ascending = renderer_->RenderPlottingTrajectoryInWorld(
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Please keep this formatted the way it was.  It will make the diff easier to review.

File: ksp_plugin/plugin.cpp [hunk 3]
```diff
@@ -937,18 +941,75 @@ void Plugin::ComputeAndRenderNodes(
...
+  std::optional<DiscreteTrajectory<Navigation>> filtered_descending_trajectory;
+
+  auto const cast_plotting_frame = dynamic_cast<
+      BodyCentredNonRotatingDynamicFrame<Barycentric, Navigation> const*>(
+      &*renderer_->GetPlottingFrame());
+  if (cast_plotting_frame != nullptr) {
+    auto const& plotting_frame = *cast_plotting_frame;
+    auto const centre = dynamic_cast_not_null<RotatingBody<Barycentric> const*>(
+        plotting_frame.centre());
+    // The plotting frame is a body-centred non-rotating frame; as this frame
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: This algorithm has nothing to do in the plugin, which is already big enough.  It should be moved elsewhere (probably `apsides.{hpp,cpp}` and properly unit-tested.

File: ksp_plugin_adapter/ksp_plugin_adapter.cs [hunk 4]
```diff
@@ -1090,8 +1090,7 @@ private enum UnityLayers {
     // Advance the lagging vessels and kill those which collided with a
     // celestial.
     {
-      DisposableIterator collided_vessels;
-      plugin_.CatchUpLaggingVessels(out collided_vessels);
+      plugin_.CatchUpLaggingVessels(out DisposableIterator collided_vessels);
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: This is a perfectly valid cleanup, but I'd rather avoid this stuff with mixing semantics as it makes the review really hard.  The same holds for the places where string concatenations are replaced by `$@""` strings.

Could you move all the cleanups to a different PR that would go in first?

File: ksp_plugin_adapter/map_node_pool.cs [hunk 5]
```diff
@@ -45,30 +45,79 @@ public enum NodeSource {
...
+        type == MapObject.ObjectType.Periapsis) {
+      CelestialBody fixed_body = reference_frame.selected_celestial;
+      associated_map_object = fixed_body.MapObject;
+      colour = fixed_body.orbit == null
+            ? XKCDColors.SunshineYellow
+            : fixed_body.orbitDriver.Renderer.nodeColor;
+    } else if (type == MapObject.ObjectType.ApproachIntersect) {
+      associated_map_object = reference_frame.target_override.mapObject;
+      colour = XKCDColors.Chartreuse;
+    } else {
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: `else if`

File: ksp_plugin_adapter/map_node_pool.cs [hunk 6]
```diff
@@ -111,132 +160,105 @@ public enum NodeSource {
...
+                    Culture.culture);
+            caption.captionLine2 = $"{speed:N0} m/s".ToString(Culture.culture);
           } else if (properties_[node].object_type ==
                          MapObject.ObjectType.AscendingNode ||
                      properties_[node].object_type ==
                          MapObject.ObjectType.DescendingNode) {
-            String name = properties_[node].object_type ==
-                                  MapObject.ObjectType.AscendingNode
-                              ? "Ascending Node"
-                              : "Descending Node";
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: Gratuitous formatting change.

File: ksp_plugin/plugin.cpp (hunk 1 above)
Comment (eggrobin) [reply ksp_plugin/plugin.cpp]: I `dynamic_cast_not_null<OblateBody<Barycentric> const*>(centre)` below, and that becomes `&dynamic_cast<OblateBody<Barycentric> const&>(centre)`, which, being passed as a `not_null` parameter, induces a gratuitous nullness check.
Do we want that?

Comment (pleroy) [reply ksp_plugin/plugin.cpp]: Ah, I missed that.

File: ksp_plugin_adapter/map_node_pool.cs [hunk 7]
```diff
@@ -45,30 +45,79 @@ public enum NodeSource {
...
-                            Vessel vessel,
-                            CelestialBody celestial) {
+                            ReferenceFrameSelector reference_frame) {
     // We render at most 64 markers of one type and one provenance (e.g., at
     // most 64 perilunes for the prediction of the active vessel).  This is
     // more than is readable, and keeps the size of the map node pool under
     // control.
+    MapObject associated_map_object;
+    UnityEngine.Color colour;
+    if (type == MapObject.ObjectType.Apoapsis ||
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: `switch`?

File: ksp_plugin_adapter/reference_frame_selector.cs [hunk 8]
```diff
@@ -221,6 +221,21 @@ public enum FrameType {
...
+         frame_type == FrameType.BODY_SURFACE)) {
+      return $"equator of {selected_celestial.NameWithArticle()}";
+    }
+    string secondary = target_override != null
+        ? "the target"
+        : selected_celestial.NameWithArticle();
+    string primary = target_override
+        ? selected_celestial.NameWithArticle()
+        : selected_celestial.referenceBody.NameWithArticle();
+   return $"orbit of {secondary} around {primary}";
```
Comment (pleroy) [inline ksp_plugin_adapter/reference_frame_selector.cs]: Bad indent.

File: ksp_plugin_adapter/map_node_pool.cs (hunk 6 above)
Comment (eggrobin) [reply ksp_plugin_adapter/map_node_pool.cs]: OBE.

File: ksp_plugin_adapter/map_node_pool.cs (hunk 5 above)
Comment (eggrobin) [reply ksp_plugin_adapter/map_node_pool.cs]: OBE.

File: ksp_plugin_adapter/ksp_plugin_adapter.cs (hunk 4 above)
Comment (eggrobin) [reply ksp_plugin_adapter/ksp_plugin_adapter.cs]: Done in #2172.

File: ksp_plugin/plugin.cpp (hunk 3 above)
Comment (eggrobin) [reply ksp_plugin/plugin.cpp]: Done.

File: ksp_plugin/plugin.cpp [hunk 9]
```diff
@@ -929,14 +930,34 @@ void Plugin::ComputeAndRenderNodes(
     std::unique_ptr<DiscreteTrajectory<World>>& descending) const {
   auto const trajectory_in_plotting =
       renderer_->RenderBarycentricTrajectoryInPlotting(begin, end);
+
+  auto const cast_plotting_frame = dynamic_cast<
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: `auto const* const`

File: ksp_plugin/plugin.cpp [hunk 10]
```diff
@@ -929,14 +930,34 @@ void Plugin::ComputeAndRenderNodes(
...
+      BodyCentredNonRotatingDynamicFrame<Barycentric, Navigation> const*>(
+      &*renderer_->GetPlottingFrame());
+  // The body-centred non rotating frame does not rotate; its reference plane is
+  // not an inherent dynamical property.   When using this frame, discard the
```
Comment (pleroy) [inline ksp_plugin/plugin.cpp]: Remove extra space.

File: ksp_plugin_adapter/ksp_plugin_adapter.cs [hunk 11]
```diff
@@ -1864,6 +1864,8 @@ private enum UnityLayers {
 
   private void RenderPredictionMarkers(String vessel_guid,
                                        XYZ sun_world_position) {
+    string reference_plane =
+        plotting_frame_selector_.ReferencePlaneDescription();
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: Remove.

File: ksp_plugin_adapter/ksp_plugin_adapter.cs [hunk 12]
```diff
@@ -1906,45 +1905,38 @@ private enum UnityLayers {
...
+          MapObject.ObjectType.DescendingNode,
+          MapNodePool.NodeSource.PREDICTION,
+          plotting_frame_selector_);
     }
   }
 
   private void RenderFlightPlanMarkers(String vessel_guid,
                                        XYZ sun_world_position) {
+    string reference_plane =
+        plotting_frame_selector_.ReferencePlaneDescription();
```
Comment (pleroy) [inline ksp_plugin_adapter/ksp_plugin_adapter.cs]: Remove.

File: ksp_plugin_adapter/map_node_pool.cs [hunk 13]
```diff
@@ -45,30 +45,86 @@ public enum NodeSource {
...
         nodes_.Add(MakePoolNode());
-      } else if (properties_[nodes_[pool_index_]].object_type != type ||
-                 properties_[nodes_[pool_index_]].celestial != celestial) {
+      } else if (properties_[nodes_[pool_index_]].object_type !=
+                     node_properties.object_type ||
+                 properties_[nodes_[pool_index_]].colour !=
+                     node_properties.colour) {
         // KSP attaches labels to its map nodes, but never detaches them.
         // If the node changes type, we end up with an arbitrary combination of
         // labels Ap, Pe, AN, DN.
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: Change "celestial" on next line.

File: ksp_plugin_adapter/map_node_pool.cs [hunk 14]
```diff
@@ -112,62 +168,32 @@ public enum NodeSource {
...
     new_node.OnUpdateType +=
         (KSP.UI.Screens.Mapview.MapNode node,
          KSP.UI.Screens.Mapview.MapNode.TypeData type) => {
           var properties = properties_[node];
-          if (properties.object_type == MapObject.ObjectType.Periapsis &&
-              properties.celestial.GetAltitude(properties.world_position) < 0) {
-            type.oType = MapObject.ObjectType.PatchTransition;
+          type.oType = properties.object_type;
+          if (properties.object_type ==
+              MapObject.ObjectType.PatchTransition) {
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: Would fit on previous line.

File: ksp_plugin_adapter/map_node_pool.cs [hunk 15]
```diff
@@ -192,13 +218,13 @@ public enum NodeSource {
                   properties.object_type == MapObject.ObjectType.Periapsis
                       ? "Periapsis"
                       : "Apoapsis";
-              CelestialBody celestial = properties.celestial;
+              CelestialBody celestial =
+                  properties.reference_frame.selected_celestial;
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: Would fit on previous line?

File: ksp_plugin_adapter/map_node_pool.cs [hunk 16]
```diff
@@ -210,37 +236,38 @@ public enum NodeSource {
...
-                     separation:N0} m</color>".ToString(Culture.culture);
+                  $@"{source} Target Approach : {separation:N0} m".ToString(
+                      Culture.culture);
               caption.captionLine2 =
                   $"{speed:N0} m/s".ToString(Culture.culture);
               break;
             }
+            case MapObject.ObjectType.PatchTransition: {
+              CelestialBody celestial =
+                  properties.reference_frame.selected_celestial;
```
Comment (pleroy) [inline ksp_plugin_adapter/map_node_pool.cs]: Would fit on previous line?

File: ksp_plugin_adapter/reference_frame_selector.cs [hunk 17]
```diff
@@ -221,6 +221,21 @@ public enum FrameType {
...
+  public String ReferencePlaneDescription() {
+    if (!target_override &&
+        (frame_type == FrameType.BODY_CENTRED_NON_ROTATING ||
+         frame_type == FrameType.BODY_SURFACE)) {
+      return $"equator of {selected_celestial.NameWithArticle()}";
+    }
+    string secondary = target_override != null
+        ? "the target"
+        : selected_celestial.NameWithArticle();
+    string primary = target_override
```
Comment (pleroy) [inline ksp_plugin_adapter/reference_frame_selector.cs]: Compare `target_override`.

File: physics/apsides.hpp [hunk 18]
```diff
@@ -23,12 +28,14 @@ void ComputeApsides(Trajectory<Frame> const& reference,
 // and |end| with the xy plane.  Appends the crossings that go towards the
 // |north| side of the xy plane to |ascending|, and those that go away from the
 // |north| side to |descending|.
-template<typename Frame>
+/// Nodes for which |filter| returns false are excluded.
```
Comment (pleroy) [inline physics/apsides.hpp]: Too many slashes.  Also, `predicate` instead of `filter`.

File: ksp_plugin_adapter/map_node_pool.cs (hunk 15 above)
Comment (eggrobin) [reply ksp_plugin_adapter/map_node_pool.cs]: No.

File: ksp_plugin_adapter/map_node_pool.cs (hunk 16 above)
Comment (eggrobin) [reply ksp_plugin_adapter/map_node_pool.cs]: No.
